numpy/lib/_function_base_impl.py

```diff
@@ -4535,8 +4535,26 @@ def quantile(a,
         if axis is not None:
             axis = _nx.normalize_axis_tuple(axis, a.ndim, argname="axis")
         weights = _weights_are_valid(weights=weights, a=a, axis=axis)
+        if weights.dtype != object:
+            if np.any(np.isinf(weights)):
```

Comment:
Another general comment: as written, the common case has to go through a lot of checks. I think it would be better to optimize for the common case, and not worry too much about distinguishing failure cases. E.g., you can do just one evaluation with:
```
if not np.all(np.isfinite(weights)):
    raise ValueError("weights must be finite.")
```